### ISSUE TITLE
Change transformer-base and transformer-big aliases to use preprocess normalisation

### ISSUE THREAD

**Issue by XapaJIaMnu** [marian-nmt/marian-dev#715]
Hey,

A number of papers have  surfaced recently that show that one of the reasons why transformers are hard to train is that the original Vaswani et al paper described postprocessing layer normalisation, while the code in fact used preprocessing layer normalisation. Examples:
https://twitter.com/srchvrs/status/1301933527391514624
https://www.aclweb.org/anthology/2020.acl-main.38.pdf
https://slideslive.com/38929843/transformers-without-tears-improving-the-normalization-of-selfattention
I experienced the issue myself, when I tried training some low resource scenario and discovered that pre-norm dramatically improved my performance.

I think it is worth to change the default setting in aliases, as this is somewhat obscure bug/feature of transformers and may lead to suboptimal marian experience for novice neural network users.

Proposed change. Change [here](https://github.com/marian-nmt/marian-dev/blob/master/src/common/aliases.cpp#L83) and [here](https://github.com/marian-nmt/marian-dev/blob/master/src/common/aliases.cpp#L121) to:
```yml
transformer-preprocess: "n"
transformer-postprocess: "da"
```

Cheers,

Nick

**Comment by emjotde** [comment]
The weird thing is I was never able to confirm that and usually have better results with the current setting. The pre-norm also seems to result in way worse stability in low/mixed precision scenarios (as you saw) in combination with certain dropout settings. So I am on the fence for this one.  

Could you check please what the reference implementation are doing exactly. Especially if there are any normalization steps before the output layer even if there is pre-norm.

**Comment by XapaJIaMnu** [comment]
@emjotde in addition to what @yuyang-huang said #699
https://github.com/tensorflow/tensor2tensor/blob/95d021477272c10af15cd62f25b595ad16ad514e/tensor2tensor/models/transformer.py#L1845

```python
    # if normalization is done in layer_preprocess, then it should also be done
    # on the output, since the output can grow very large, being the sum of
    # a whole stack of unnormalized layer outputs.
```
https://github.com/tensorflow/tensor2tensor/blob/95d021477272c10af15cd62f25b595ad16ad514e/tensor2tensor/models/transformer.py#L1716

Tracing back further commits I found those comments:
https://github.com/tensorflow/tensor2tensor/commit/f5c9b17e617ea9179b7d84d36b1e8162cb369f25#diff-4e58a582cf11ca649e76b4362d69e405R78
```python
      # Sequences of operations to perform on layer input and layer output.
      # Used by common_layers.layer_preprocess, common_layers.layer_postprocess
      # Each character repsesnts an operation:
      #   d: apply dropout
      #   n: apply normalization (see norm_type and norm_epsilon)
      #   a: add layer input (residual connection - only during postprocess)
      # TODO(noam): The current settings ("", "dan") are the published version
      # of the transformer.  ("n", "da") seems better for harder-to-learn
      # models, so it should probably be the default.
```
It's getting a bit confusing to hunt through the code, but there seems to be a `layer_prepostprocess` data takes care of these normalisation specifics.

**Comment by emjotde** [comment]
Cool. Thank you. I am not sure I want to change the defaults, but adding two new aliases like `transformer-base-prenorm` and `transformer-big-prenorm`  should be alright?

**Comment by XapaJIaMnu** [comment]
Yes, I imagine this would be good, coupled with some explanation added to the website/examples. When proper investigation is performed, we could tweak the tutorials more.

Also, how do you feel about a warning message when using prenorm and tensorcores/float16 code, so that we warn new users should they encounter NaNs?

**Comment by emjotde** [comment]
The NANs maybe a consequence of the missing final postnorm. I am add code for that now.

**Comment by emjotde** [comment]
@XapaJIaMnu published a first untested draft in #719 with `--task transformer-base-prenorm` you should get the desired behavior. This corresponds to `--task transformer-base --transformer-preprocess n --transformer-postprocess da --transformer-postprocess-top n`. I will start testing this now.

**Comment by emjotde** [comment]
Closing this with PR #719

**Comment by tenebrius** [comment]
I get gibberish translation when using transform-prenorm. Anybody else experiencing this?

**Comment by m-resta** [comment]
> I get gibberish translation when using transform-prenorm. Anybody else experiencing this?

I am experiencing the same problem despite the validation perplexity score is decreasing.

 My settings are:

```

    --devices 0 1 2 3 \
    --type transformer \
    --model ${MODEL_HOME_DIR}/model_bktrans-m4_ite3_scale-efficient.npz \
    --train-sets ${DATASET_DIR}/corpus.src ${DATASET_DIR}/corpus.tgt \
    --vocabs ${MODEL_HOME_DIR}/vocab.multi.spm ${MODEL_HOME_DIR}/vocab.multi.spm \
    --valid-sets ${DATASET_DIR}/devset/devset.src ${DATASET_DIR}/devset/devset.tgt \
    --dim-vocabs 36000 36000 \
    --sentencepiece-options '--character_coverage=1.0 --model_type=bpe --train_extremely_large_corpus --normalization_rule_name=nmt_nfkc' \
    --mini-batch-fit -w 34000 --fp16 \
    --sentencepiece-max-lines 2000000000 \
    --layer-normalization --tied-embeddings-src \
    --early-stopping 30 --max-length 220 \
    --valid-freq 20000  --save-freq 10000  --disp-freq 500 \
    --cost-type ce-mean --valid-metrics perplexity \
    --log ${MODEL_HOME_DIR}/train.log \
    --valid-log ${MODEL_HOME_DIR}/valid.log \
    --tempdir ${MODEL_HOME_DIR}/temp_files \
    --overwrite --keep-best \
    --seed 1111 --exponential-smoothing \
    --normalize=0.6 --beam-size=5 --quiet-translation \
    --learn-rate 0.00005 --lr-warmup 128000 --lr-decay-inv-sqrt 256000 --lr-report \
    --enc-depth 24 --dec-depth 24 \
    --no-restore-corpus \
    --shuffle-in-ram \
    --optimizer adam \
    --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \
    --transformer-dim-ffn 2048 --transformer-heads 8 \
    --transformer-dropout-attention 0.3 --transformer-dropout-ffn 0.3 \
    --dim-emb 512 --sync-sgd \
    --transformer-postprocess-emb d \
    --transformer-postprocess da \
    --transformer-preprocess n \
    --lr-warmup-at-reload \
    --transformer-dropout 0.3 --label-smoothing 0.1 \
  ```

I already trained a post-norm transformer on the same data and get good translations.
The pre-norm transformer settings are those suggested in the paper [Scale Efficiently: Insights from Pre-training and Fine-tuning Transformers](https://arxiv.org/abs/2109.10686) so the model should be roughly equivalent to  a transformer base.
